### PR TITLE
Pass parameter "transaction" to the Parse Server

### PR DESCRIPTION
**Pull request by mfkenson**
https://github.com/parse-community/Parse-SDK-JS/issues/922

Did I miss anything?

### CODE REVIEW

File: src/RESTController.js [hunk 1]
```diff
@@ -230,6 +232,11 @@ const RESTController = {
       }
     }
 
+    const transaction = options.transaction;
+    if (transaction) {
+        payload.transaction = true;
```
Comment (JeromeDeLeon) [inline src/RESTController.js]: You might want to lint your code.

Comment (mfkenson) [reply src/RESTController.js]: I am quite new to this and don't know what needs to be done exactly. Can you give me some advises how to get started?

Comment (JeromeDeLeon) [reply src/RESTController.js]: Just run `npm run lint` or `yarn lint` If you are using yarn

**Comment by davimacedo** [comment]
Nice! Thanks for contributing. Could you please add a test case?

File: src/RESTController.js (hunk 1 above)
Comment (dplewis) [reply src/RESTController.js]: Check out the [Contributing Guide](https://github.com/parse-community/Parse-SDK-JS/blob/master/CONTRIBUTING.md) for all your testing and linting needs.

**Comment by dplewis** [comment]
@davimacedo Can you take this over and add the test cases? You know more about how transactions in parse work better than anyone.

**Comment by davimacedo** [comment]
Sure. I will work on this.

**Comment by dplewis** [comment]
@davimacedo Any updates?

**Comment by davimacedo** [comment]
Sorry for the delay. I am working on this now.

**Comment by davimacedo** [comment]
@dplewis @mfkenson I started worked on the tests here but there is a problem with this implementation. In the case you have more objects to be saved/destroyed than the batch size, the SDK will split it in more than one batch and it will be generated two transactions in the server side. I think we can solve it using two different approaches:
1. Change the sdk to never split into batches when the transaction option is set to true
2. Throw an error message when the number of objects to be saved/destroyed is bigger than the batch size and the transaction option is set to true.

Thoughts?

**Comment by dplewis** [comment]
@davimacedo I think that would be a great approach. I tried that [here](https://github.com/parse-community/Parse-SDK-JS/pull/1047/files) and ran into an issue. The issue is that we wouldn't be be able to find cycles in `SaveAll`. I'm not sure the backend is setup to catch cycles.

**Comment by davimacedo** [comment]
I've just checked this problem and it will be very hard to be solved now. The not saved dependent objects need to go in a first batch because there is no warranty they will be executed in first place in the server side if we send them all together with the objects that depend on them. And if we have more batched sent to the server separately we cannot ensure the transaction. So, to simplify the solution and have a first version of this feature in the next release I suggest we simply do two verifications:

- Verify if there is any not saved dependent object when transaction=true and throw an error in the case they exist.
- Verify if there are more objects to be saved than the batch size and throw an error in the case there are.

Do you agree @dplewis ?

**Comment by dplewis** [comment]
Thats the only solution I can come up with. We will have to document it well.

**Comment by andreasanta** [comment]
@mfkenson Can you please grant me access to this PR so I can fix linting and maybe try and perform some tests?

@davimacedo Any suggestions on how to start writing tests for this? :)

**Comment by davimacedo** [comment]
Hi guys. I am working on this (to include the validations and the tests) but I found out a bug in Parse Server when deleting objects in a transaction. I am working now on a first PR to Parse Server before fixing this one here.

**Comment by santiagosemhan** [comment]
+1

**Comment by andreasanta** [comment]
@davimacedo Thanks. Please let me know whether I can be of assistance.

**Comment by vahidalizad** [comment]
Hi guys, Great work so far
I really had to use this feature so I came with the following code:
```
function getServerUrlPath() {
  let serverUrl = SERVER_URL;
  if (serverUrl[serverUrl.length - 1] !== '/') {
    serverUrl += '/';
  }
  const url = serverUrl.replace(/https?:\/\//, '');
  return url.substr(url.indexOf('/'));
}

const MyCustom = Parse.Object.extend('MyCustom');
let obj1 = new MyCustom({name: 'just a name1'});
let obj2 = new MyCustom({name: 'just a name2'});

let batch = [obj1, obj2];
let options = {};

Parse._request(
  'POST',
  'batch',
  {
    requests: batch.map(obj => {
      const params = obj._getSaveParams();
      params.path = getServerUrlPath() + params.path;
      return params;
    }),
    transaction: true,
  },
  options
)
  .then(res => console.log('success', res))
  .catch(e => console.log('failed', e));
```
But there is something I don't understand when it fails, it sends up to `REQUEST_ATTEMPT_LIMIT` retries request but when I save only one file with `await obj1.save()` it only does it once and log it once. maybe this is because the batch request error does not return `Parse.Error` and the result has a code above 500.
Is this intended and I have to change REQUEST_ATTEMPT_LIMIT somehow?

**Comment by davimacedo** [comment]
Maybe. You can set it via `CoreManager.set('REQUEST_ATTEMPT_LIMIT', retry);`

**Comment by vahidalizad** [comment]
`Parse.CoreManager.set('REQUEST_ATTEMPT_LIMIT', 1);` changes every request retry times to 1 and it works, but there is no way to set a retry value for a single request.
but thanks anyway it works.
Maybe you could change the `src/ParseServerRESTController.js` in the definition of `ParseServerRESTController` like bellow to return the error like the normal batch requests

```
if (path === "/batch") {
    let initialPromise = Promise.resolve();
    if (data.transaction === true) {
    initialPromise = config.database.createTransactionalSession();
    }
    return initialPromise.then(() => {
    const promises = data.requests.map((request) => {
        return handleRequest(
        request.method,
        request.path,
        request.body,
        options,
        config
        ).then(
        (response) => {
            return Promise.resolve({ success: response });
        },
        (error) => {
            return Promise.resolve({
            error: { code: error.code, error: error.message },
            });
        }
        );
    });
    return Promise.all(promises)
        .catch((error) => {
        if (data.transaction === true) {
            return config.database.abortTransactionalSession().then(() => {
            throw error;
            });
        } else {
            throw error;
        }
        })
        .then((result) => {
        if (data.transaction === true) {
            let error = result.find((t) => t.error && t.code);
            if (error) {
            return config.database.abortTransactionalSession().then(() => {
                return error;
            });
            }
            return config.database.commitTransactionalSession().then(() => {
            return result;
            });
        } else {
            return result;
        }
        });
    });
}
```

**Comment by davimacedo** [comment]
That makes sense. Would you be willed to continue this PR?

**Comment by vahidalizad** [comment]
> 
> 
> That makes sense. Would you be willed to continue this PR?

It's not about this PR, I have to create a pr on parse-server project, I create a new fork and when it's finished I notify you.
After I finished that I'm gonna join you in this one if you haven't finished it yet.

**Comment by RahulLanjewar93** [comment]
@davimacedo Hi guys, any updates on this?

**Comment by araskinwhichbnb** [comment]
Its Sep 2023 and there is an ominous comment above that says someone deleted this feature. Does that mean we are not going to get Transactional Integrity in the JS SDK?
@davimacedo

**Comment by RahulLanjewar93** [comment]
@mfkenson just wanted to know if there's any other pr/issue for reference since you closed this one.

**Comment by Llorx** [comment]
@mfkenson ???

**Comment by Llorx** [comment]
> Hi guys. I am working on this (to include the validations and the tests) but I found out a bug in Parse Server when deleting objects in a transaction. I am working now on a first PR to Parse Server before fixing this one here.

@davimacedo Hey, have you fixed this problem? Can I create a new PR with the transaction: true flag now?

**Comment by vahidalizad** [comment]
> That makes sense. Would you be willed to continue this PR?

@davimacedo Thank you for your patience! I apologize for the delay. it took me three years to get back to this 😅. I was surprised that this hadn't been completed yet, so I went ahead and created a PR to address it.

Here’s the link to the PR: [#2265](https://github.com/parse-community/Parse-SDK-JS/pull/2265).

**Pull request closed without merging**